### PR TITLE
fix:  Ecdsa Malleability Bug

### PR DESCRIPTION
**Pull request by suyash67**
# Description

ECDSA has an inherent malleability because of the way $s$ in an ECDSA signature is computed. Read more about this [here](https://www.derpturkey.com/inherent-malleability-of-ecdsa-signatures/amp/). This PR adds the constraint $s < |Fr|/2$ in ECDSA verification to fix the malleability. Additionally, in the signature construction, we make sure that $s < |Fr|/2$.

Acknowledgements: @Rumata888 corrected the constraint on $s$ in verification.

# Checklist:

- [x] I have reviewed my diff in github, line by line.
- [x] Every change is related to the PR description.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to the issue(s) that it resolves.
- [x] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [x] There are no circuit changes, OR specifications in `/markdown/specs` have been updated.
- [x] There are no circuit changes, OR a cryptographer has been assigned for review.
- [x] I've updated any terraform that needs updating (e.g. environment variables) for deployment.
- [x] The branch has been rebased against the head of its merge target.
- [x] I'm happy for the PR to be merged at the reviewer's next convenience.
- [x] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [x] If existing code has been modified, such documentation has been added or updated.

### CODE REVIEW

File: cpp/src/barretenberg/stdlib/primitives/bigfield/bigfield_impl.hpp [hunk 1]
```diff
@@ -1664,6 +1664,67 @@ template <typename C, typename T> void bigfield<C, T>::assert_is_in_field() cons
     }
 }
 
+template <typename C, typename T> void bigfield<C, T>::assert_less_than(const uint256_t& upper_limit) const
```
Comment (Rumata888) [inline cpp/src/barretenberg/stdlib/primitives/bigfield/bigfield_impl.hpp]: Please create a TODO to merge it with assert_is_in_field in the future

File: cpp/src/barretenberg/stdlib/primitives/bigfield/bigfield.test.cpp [hunk 2]
```diff
@@ -712,6 +712,42 @@ template <typename Composer> class stdlib_bigfield : public testing::Test {
         EXPECT_EQ(result, true);
     }
 
+    static void test_assert_less_than_success()
```
Comment (Rumata888) [inline cpp/src/barretenberg/stdlib/primitives/bigfield/bigfield.test.cpp:715]: Please create a test that checks specifically for the edge case a<a. I think that the assert_less_than function will allow it, since we do a-a and check that it can be range constrained.

File: cpp/src/barretenberg/stdlib/primitives/bigfield/bigfield.hpp [hunk 3]
```diff
@@ -248,6 +248,7 @@ template <typename Composer, typename T> class bigfield {
     bigfield conditional_select(const bigfield& other, const bool_t<Composer>& predicate) const;
 
     void assert_is_in_field() const;
+    void assert_less_than(const uint256_t& upper_limit) const;
```
Comment (Rumata888) [inline cpp/src/barretenberg/stdlib/primitives/bigfield/bigfield.hpp]: Why should it be a reference?

**Review by Rumata888 — APPROVED** [review]
I've made two changes:
1) The function was actually less or equal.
2) Once I changed it to be less_than, the logic for constraining s became incorrect. You were checking that s < (Fr::modulus / 2). It is different from 2*s < Fr::modulus, because Fr::modulus is odd. Think  mod 3. If you check that something is less than 3/2, then the only option is zero, while 1 should also be a legitimate value.

File: cpp/src/barretenberg/stdlib/primitives/bigfield/bigfield.test.cpp [hunk 4]
```diff
@@ -712,6 +712,53 @@ template <typename Composer> class stdlib_bigfield : public testing::Test {
...
+            // fq_ct c = a + a + a + a - b - b - b - b;
+            c.assert_less_than(bit_mask + 1);
+            uint256_t result = (c.get_value().lo);
+            EXPECT_EQ(result, uint256_t(expected));
+            EXPECT_EQ(c.get_value().get_msb() < num_bits, true);
+        }
+        bool result = composer.check_circuit();
+        EXPECT_EQ(result, true);
+        // Checking edge conditions
+        fq random_input = fq::random_element();
```
Comment (kevaundray) [inline cpp/src/barretenberg/stdlib/primitives/bigfield/bigfield.test.cpp:746]: nit, but not a blocker: unit tests should ideally not contain random elements as they then become unreproducable

**Review by kevaundray — APPROVED** [review]
Looks good to me

File: cpp/src/barretenberg/stdlib/primitives/bigfield/bigfield.test.cpp (hunk 4 above)
Comment (Rumata888) [reply cpp/src/barretenberg/stdlib/primitives/bigfield/bigfield.test.cpp:746]: You are partially right. I'd add something like "random_element_for_test" into the field, so that it would automatically print it. But it is actually better to have random elements than static, because if there is a bug there is a chance that it triggers. And then you can stress the test to find it (that was the case with construct_addition_chains)

**Comment by kevaundray** [comment]
Since this modifies the stdlib by adding extra constraints, I'm marking this a breaking change

**Comment by kevaundray** [comment]
Update: It was decided that changes to the stdlib will not be marked as breaking changes as this would include most commits

**Comment by suyash67** [comment]
> I've made two changes:
> 
> 1. The function was actually less or equal.
> 2. Once I changed it to be less_than, the logic for constraining s became incorrect. You were checking that s < (Fr::modulus / 2). It is different from 2*s < Fr::modulus, because Fr::modulus is odd. Think  mod 3. If you check that something is less than 3/2, then the only option is zero, while 1 should also be a legitimate value.

Nice catch, thanks Kesha!

File: cpp/src/barretenberg/stdlib/primitives/bigfield/bigfield.hpp (hunk 3 above)
Comment (suyash67) [reply cpp/src/barretenberg/stdlib/primitives/bigfield/bigfield.hpp]: It doesn't need to be, you're right. Removing it.
